src/com/linkedin/parseq/FusionTraceContext.java

```diff
@@ -1,23 +1,55 @@
 package com.linkedin.parseq;
 
+import com.linkedin.parseq.internal.IdGenerator;
 import com.linkedin.parseq.trace.ShallowTraceBuilder;
 
 
-public class FusionTraceContext {
+/**
+ * FusionTraceContext encapsulates information necessary to generate trace
+ * of fusion tasks.
+ *
+ * This is internal class and it should not be extended or used outside ParSeq.
+ * It may change or be removed at any time in a backwards incompatible way.
+ *
+ * @author Jaroslaw Odzga (jodzga@linkedin.com)
+ */
+class FusionTraceContext {
 
-  private final ShallowTraceBuilder _trigger;
-  private final Context _context;
+  /* Task that initiated propagation */
+  private final ShallowTraceBuilder _propagationInitiator;
 
-  public FusionTraceContext(Context context, ShallowTraceBuilder trigger) {
-    _context = context;
-    _trigger = trigger;
+  /* Context of a parent task, it will be different than propagationInitiator if fusion is applied on async task */
+  private final Context _parent;
+
+  /* Surrogate trace is a trace of last transformation in a chain - outermost propagator */
+  private ShallowTraceBuilder _surrogate;
+
+  /* Description of the last transformation in the chain */
+  private final String _desc;
+
+  public FusionTraceContext(Context parent, ShallowTraceBuilder propagationInitiator, String desc) {
+    _parent = parent;
+    _propagationInitiator = propagationInitiator;
+    _desc = desc;
+  }
+
+  public ShallowTraceBuilder getPropagationInitiator() {
+    return _propagationInitiator;
+  }
+
+  public Context getParent() {
+    return _parent;
   }
 
-  public ShallowTraceBuilder getTrigger() {
-    return _trigger;
+  public ShallowTraceBuilder getSurrogate() {
+    return _surrogate;
   }
 
-  public Context getContext() {
-    return _context;
+  public void createSurrogate() {
```

Comment:
Because surrogate is set only when there are at least 2 transformation. This complication is necessary to avoid wrapping single transformation with "fuse" parent.